### ISSUE TITLE
op/alarm.t hangs on 5.11.0 (Windows Vista only)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9251]
Migrated from [rt.perl.org#51674](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=51674) (status was 'resolved')

Searchable as RT51674$

**Comment by p5pRT** [comment]
### From @sisyphus

To&#8203;: perlbug@&#8203;perl\.org
Subject&#8203;: op/alarm\.t hangs on 5\.11\.0 \(Windows Vista only\)
Reply\-To&#8203;: sisyphus1@&#8203;optusnet\.com\.au
Message\-Id&#8203;: \<5\.11\.0\_3628\_1205376264@&#8203;desktop2>

This is a bug report for perl from sisyphus1@&#8203;optusnet\.com\.au\,
generated with the help of perlbug 1\.36 running under perl 5\.11\.0\.

Firstly this problem affects Windows Vista only\. At least\, I've checked on 
Windows XP and there's no problem\.
Secondly\, it affects only perl 5\.11\.0 \(March 10 snapshot\)\. All works fine on 
5\.10\.0\.

The test script \(based upon the problem code in op/alarm\.t\) :



---
use warnings;
use strict;

my $start\_time = time;
local $SIG\{ALRM\} = sub \{ die "ALARM\! "\, time \- $start\_time\, "\\n" \};
alarm 3;
print "Got to here\\n";
1 while \(time \- $start\_time \< 6\);


---
On Windows Vista\, 5\.11\.0 that outputs the "Got to here\\n" and then hangs\.

There is still no additional output even if I rewrite the while\(\) loop as&#8203;:

print "\." while \(time \- $start\_time \< 6\);

I can\, however\, coerce some additional output if I rewrite the while loop 
as&#8203;:

print "\." while 1;

That will then print out 4096 dots to the screen \(in a fraction of a second\) 
before hanging\.

I'm running Vista64 on an AMD64 box\. I don't expect Vista32 behaves an 
differently\.

Cheers\,
Rob\-\-\-
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl 5\.11\.0&#8203;:

Configured by Rob at Tue Mar 11 09&#8203;:51&#8203;:36 2008\.

Summary of my perl5 \(revision 5 version 11 subversion 0 patchlevel 33458\) 
configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=6\.0\, archname=MSWin32\-x64\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags 
' \-s \-O2 \-DWIN32 \-DHAVE\_DES\_FCRYPT  \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS 
 \-fno\-strict\-aliasing \-mms\-bitfields \-DPERL\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-s \-O2'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion=''\, gccversion='3\.4\.5'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='long long'\, 
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='g\+\+'\, ldflags 
='\-s \-L"c&#8203;:\\bleadperl\\lib\\CORE" \-L"C&#8203;:\\home\\rob\\mingw\_vista\\i686\-pc\-mingw32\\lib"'
&nbsp;    libpth=C&#8203;:\\home\\rob\\mingw\_vista\\i686\-pc\-mingw32\\lib
&nbsp;    libs=\-lmsvcrt \-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32 
 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr \-lwinmm 
 \-lversion \-lodbc32 \-lodbccp32
&nbsp;    perllibs=\-lmsvcrt \-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32 
 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr \-lwinmm 
 \-lversion \-lodbc32 \-lodbccp32
&nbsp;    libc=\-lmsvcrt\, so=dll\, useshrplib=true\, libperl=libperl511\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, 
lddlflags='\-mdll \-s \-L"c&#8203;:\\bleadperl\\lib\\CORE" \-L"C&#8203;:\\home\\rob\\mingw\_vista\\i686\-pc\-mingw32\\lib"'

Locally applied patches&#8203;:
&nbsp;    DEVEL



---
@&#8203;INC for perl 5\.11\.0&#8203;:
&nbsp;    C&#8203;:/bleadperl/lib
&nbsp;    C&#8203;:/bleadperl/site/lib
&nbsp;    \.



---
Environment for perl 5\.11\.0&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=C&#8203;:\\Program Files\\Windows 
NT\\Accessories;C&#8203;:\\\_32\\third\_party\_app;C&#8203;:\\GNUWin32\\bin;C&#8203;:\\bleadperl\\bin;C&#8203;:\\\_32\\dmake;C&#8203;:\\home\\rob\\mingw\_vista\\i686\-pc\-mingw32\\bin;c&#8203;:\\ruby\\bin;C&#8203;:\\Windows\\system32;C&#8203;:\\Windows;C&#8203;:\\Windows\\System32\\Wbem;C&#8203;:\\batch;C&#8203;:\\\_32\\msys\\1\.0\\bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From robertmay@cpan.org

On 13/03/2008\, via RT Sisyphus \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  "Sisyphus"
>  \# Please include the string&#8203;:  \[perl \#51674\]
>  \# in the subject line of all future correspondence about this issue\.
>  \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=51674 >
>
>  To&#8203;: perlbug@&#8203;perl\.org
>  Subject&#8203;: op/alarm\.t hangs on 5\.11\.0 \(Windows Vista only\)
>  Reply\-To&#8203;: sisyphus1@&#8203;optusnet\.com\.au
>  Message\-Id&#8203;: \<5\.11\.0\_3628\_1205376264@&#8203;desktop2>
>
>  This is a bug report for perl from sisyphus1@&#8203;optusnet\.com\.au\,
>  generated with the help of perlbug 1\.36 running under perl 5\.11\.0\.
>
>  Firstly this problem affects Windows Vista only\. At least\, I've checked on
>  Windows XP and there's no problem\.
>  Secondly\, it affects only perl 5\.11\.0 \(March 10 snapshot\)\. All works fine on
>  5\.10\.0\.
>
>  The test script \(based upon the problem code in op/alarm\.t\) :
>
>  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>  use warnings;
>  use strict;
>
>  my $start\_time = time;
>  local $SIG\{ALRM\} = sub \{ die "ALARM\! "\, time \- $start\_time\, "\\n" \};
>  alarm 3;
>  print "Got to here\\n";
>  1 while \(time \- $start\_time \< 6\);
>  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>
>  On Windows Vista\, 5\.11\.0 that outputs the "Got to here\\n" and then hangs\.
>
>  There is still no additional output even if I rewrite the while\(\) loop as&#8203;:
>
>  print "\." while \(time \- $start\_time \< 6\);
>
>  I can\, however\, coerce some additional output if I rewrite the while loop
>  as&#8203;:
>
>  print "\." while 1;
>
>  That will then print out 4096 dots to the screen \(in a fraction of a second\)
>  before hanging\.
>
>  I'm running Vista64 on an AMD64 box\. I don't expect Vista32 behaves an
>  differently\.

This is almost certainly down to change \#33255 \(my doing\)\.  I can't
immediately see a problem\, and don't have access to Vista to try it on
\(I only have Win98\, Win2K\, WinXP\, all 32 \-bit versions\)\.

I'm also about to start a move from India back to the UK\, so may not
have time to look into this further for a couple of week\, but I will
lokk into it as soon as I have time if no\-one else bets me to it\.

I'd appreciate confirmation that this also happens on the 32\-bit
version of Vista\, so that we can eliminate the 32/64\-bit issues \- I
can see some dubious casts in the code that may be throwing away
information\.

I think the most likely candidate for the problem are these lines in
win32\_async\_check \(in win32/win32\.c\)&#8203;:

2123&#8203;:     /\* Call PeekMessage\(\) to mark all pending messages in the
queue as "old"\.
2124&#8203;:      \* This is necessary when we are being called by win32\_msgwait\(\) to
2125&#8203;:      \* make sure MsgWaitForMultipleObjects\(\) stops reporting the
same waiting
2126&#8203;:      \* message over and over\.  An example how this can happen is when
2127&#8203;:      \* Perl is calling win32\_waitpid\(\) inside a GUI application
and the GUI
2128&#8203;:      \* is generating messages before the process terminated\.
2129&#8203;:      \*/
2130&#8203;:     while \(PeekMessage\(&msg\, NULL\, 0\, 0\, PM\_NOREMOVE|PM\_NOYIELD\)\)
2131&#8203;:         /\* keep going \*/ ;

Does removing the 'while' here fix the problem \(replace lines 2130 and
2131 with&#8203;:

&nbsp;  PeekMessage\(&msg\, NULL\, 0\, 0\, PM\_NOREMOVE|PM\_NOYIELD\);

Also\, can you confirm that when the process hangs it hangs at 100% CPU usage\.


We've seen problems with alarm\(\) and Windows 2003 Server and Vista
before\.  I can't see any explanation of what the difference was
between earlier OS versions and these in the list archive\, but the
problems resulted in change \#26379 \(and follow\-ups 27527 and 29868\)\,
where Jan introduced the message window concept\.   Jan\, can you shed
any light on the reason why the old thread\-message only implementation
didn't play nicely with Vista/2003?

Thanks\,
Rob\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @sisyphus


\-\-\-\-\- Original Message \-\-\-\-\- 
From&#8203;: "Robert May" \<robertmay@&#8203;cpan\.org>
\.
\.
> On 13/03/2008\, via RT Sisyphus \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> \# New Ticket Created by  "Sisyphus"
>>  \# Please include the string&#8203;:  \[perl \#51674\]
>>  \# in the subject line of all future correspondence about this issue\.
>>  \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=51674 >
>>
>>  To&#8203;: perlbug@&#8203;perl\.org
>>  Subject&#8203;: op/alarm\.t hangs on 5\.11\.0 \(Windows Vista only\)
>>  Reply\-To&#8203;: sisyphus1@&#8203;optusnet\.com\.au
>>  Message\-Id&#8203;: \<5\.11\.0\_3628\_1205376264@&#8203;desktop2>
>>
>>  This is a bug report for perl from sisyphus1@&#8203;optusnet\.com\.au\,
>>  generated with the help of perlbug 1\.36 running under perl 5\.11\.0\.
>>
>>  Firstly this problem affects Windows Vista only\. At least\, I've checked
>> on
>>  Windows XP and there's no problem\.
>>  Secondly\, it affects only perl 5\.11\.0 \(March 10 snapshot\)\. All works
>> fine on
>>  5\.10\.0\.
>>
>>  The test script \(based upon the problem code in op/alarm\.t\) :
>>
>>  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>  use warnings;
>>  use strict;
>>
>>  my $start\_time = time;
>>  local $SIG\{ALRM\} = sub \{ die "ALARM\! "\, time \- $start\_time\, "\\n" \};
>>  alarm 3;
>>  print "Got to here\\n";
>>  1 while \(time \- $start\_time \< 6\);
>>  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>
>>  On Windows Vista\, 5\.11\.0 that outputs the "Got to here\\n" and then
>> hangs\.
>>
>>  There is still no additional output even if I rewrite the while\(\) loop
>> as&#8203;:
>>
>>  print "\." while \(time \- $start\_time \< 6\);
>>
>>  I can\, however\, coerce some additional output if I rewrite the while
>> loop
>>  as&#8203;:
>>
>>  print "\." while 1;
>>
>>  That will then print out 4096 dots to the screen \(in a fraction of a
>> second\)
>>  before hanging\.
>>
>>  I'm running Vista64 on an AMD64 box\. I don't expect Vista32 behaves an
>>  differently\.
>
> This is almost certainly down to change \#33255 \(my doing\)\.  I can't
> immediately see a problem\, and don't have access to Vista to try it on
> \(I only have Win98\, Win2K\, WinXP\, all 32 \-bit versions\)\.
>
> I'm also about to start a move from India back to the UK\, so may not
> have time to look into this further for a couple of week\, but I will
> lokk into it as soon as I have time if no\-one else bets me to it\.
>
> I'd appreciate confirmation that this also happens on the 32\-bit
> version of Vista\, so that we can eliminate the 32/64\-bit issues \- I
> can see some dubious casts in the code that may be throwing away
> information\.
>
> I think the most likely candidate for the problem are these lines in
> win32\_async\_check \(in win32/win32\.c\)&#8203;:

\[snip\]

> Does removing the 'while' here fix the problem \(replace lines 2130 and
> 2131 with&#8203;:
>
>  PeekMessage\(&msg\, NULL\, 0\, 0\, PM\_NOREMOVE|PM\_NOYIELD\);
>

Seems to do the trick\. \(My mind boggles at the amount of time \*I\* could have 
spent in trying to find a solution\.\)

> Also\, can you confirm that when the process hangs it hangs at 100% CPU
> usage\.

Yep \.\.\. well \.\.\. it hangs at around 50% usage \.\.\. but it's dual core  \.\.\.\.
\(I don't understand computers very well :\-\)

\.
\.

Also similarly affected \(by\, I think\, the same bug\) were :

fork\.t
io\_mutlihomed\.t
io\_sock\.t
socketpair\.t
GetCurrentThreadId\.t

There may have been others \- I never did manage to get right through the
test suite without having to kill a hang\. \(I think GetCurrentThreadId\.t is
getting close to the end of the suite\, so there probably weren't too many
more after that \.\. if any\.\)

I now get failures with&#8203;:

lib/CPANPLUS/Dist/Build/t/02\_CPANPLUS\-Dist\-Build\.t
lib/ExtUtils/CBuilder/t/01\-basic\.t
lib/Module/Build/t/ppm\.t
lib/Module/Build/t/xs\.t

but I get the same on my XP box\, and those failures have already been 
reported in Steve Hay's smoke tests\. I took a bit of a look at what was 
happening there \- all failures seem to be a result of changes to 
ExtUtils&#8203;::CBuilder\. Should I submit a bug report about that ? \(I'm assuming 
it's unnecessary to do that\, given that Steve's tests have already reported 
the problem\.\)

There's also a puzzling failure with ext/threads/shared/t/stress\.t\. Seems 
that\, because one TODO test actually passed\, 'dmake test' feels compelled to 
report a failure\. \(One of Steve's smoke reports also reported this\.\)

And I've just noticed a failure with tests 8 and 9 of 
lib/ExtUtils/t/can\_write\_dir\.t\. I don't recall seeing a similar failure on 
my XP box\, and I don't recall seeing it mentioned in any of the smoke 
reports either\. \(I'll take a closer look at it\, and submit another bug 
report \- if warranted\.\)

Thanks Rob\. Enjoy the glorious English Spring \.\.\.\. brrrrr :\-\)
It was 39 degrees Celsius here today \.\.\. and that's Autumn \!\!

Cheers\,
Rob

**Comment by p5pRT** [comment]
### From robertmay@cpan.org

On 13/03/2008\, Sisyphus \<sisyphus1@&#8203;optusnet\.com\.au> wrote&#8203;:
>  > On 13/03/2008\, via RT Sisyphus \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>  >> \# New Ticket Created by  "Sisyphus"
>  >>  \# Please include the string&#8203;:  \[perl \#51674\]
>  >>  \# in the subject line of all future correspondence about this issue\.
>  >>  \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=51674 >
>  >>
>  >>  To&#8203;: perlbug@&#8203;perl\.org
>  >>  Subject&#8203;: op/alarm\.t hangs on 5\.11\.0 \(Windows Vista only\)
>  >>  Reply\-To&#8203;: sisyphus1@&#8203;optusnet\.com\.au
>  >>  Message\-Id&#8203;: \<5\.11\.0\_3628\_1205376264@&#8203;desktop2>

\[snip\]

>  > I think the most likely candidate for the problem are these lines in
>  > win32\_async\_check \(in win32/win32\.c\)&#8203;:
>
> \[snip\]
>
>  > Does removing the 'while' here fix the problem \(replace lines 2130 and
>  > 2131 with&#8203;:
>  >
>  >  PeekMessage\(&msg\, NULL\, 0\, 0\, PM\_NOREMOVE|PM\_NOYIELD\);
>  >
>
> Seems to do the trick\. \(My mind boggles at the amount of time \*I\* could have
>  spent in trying to find a solution\.\)

Cool\. I'll post a patch to P5P\.

>  > Also\, can you confirm that when the process hangs it hangs at 100% CPU
>  > usage\.
>
> Yep \.\.\. well \.\.\. it hangs at around 50% usage \.\.\. but it's dual core  \.\.\.\.
>  \(I don't understand computers very well :\-\)
>

Thanks \- that's consistent with a single thread running at 100% on one
of the cores\.  It's a pretty good indication that we're stuck in a
tight loop somewhere\.

Rob\.

**Comment by p5pRT** [comment]
### From rob@themayfamily.me.uk

A problem with my changes to give win32 proper messages loops \(applied
as change \#33255\) has been identified on Vista\.

The patch below fixes this \(patch against blead@&#8203;33505\)\.  I haven't
checked it out on other win32 OS\, but is closer to what was in
win32\_async\_check\(\) before I got my hands on it\.  I can't claim to
understand the difference in behaviour on Vista to other OS\, but the
proposed change is more consistent with the MS documentation for
PeekMessage\(\)\.

I believe that with this change bug \#51674 can be closed\.

Regards\,
Rob\.

<details><summary>Inline Patch</summary>

```diff
--- win32.c.orig	Mon Feb 25 23:17:32 2008
+++ win32.c	Thu Mar 13 19:16:22 2008
@@ -2127,8 +2127,7 @@ win32_async_check(pTHX)
      * Perl is calling win32_waitpid() inside a GUI application and the GUI
      * is generating messages before the process terminated.
      */
-    while (PeekMessage(&msg, NULL, 0, 0, PM_NOREMOVE|PM_NOYIELD))
-        /* keep going */ ;
+    PeekMessage(&msg, NULL, 0, 0, PM_NOREMOVE|PM_NOYIELD);

     /* Above or other stuff may have set a signal flag */
     if (PL_sig_pending)
```
</details>

**Comment by p5pRT** [comment]
### From rob@themayfamily.me.uk

<details><summary>win32.c.patch</summary>

```diff
--- win32.c.orig	Mon Feb 25 23:17:32 2008
+++ win32.c	Thu Mar 13 19:16:22 2008
@@ -2127,8 +2127,7 @@ win32_async_check(pTHX)
      * Perl is calling win32_waitpid() inside a GUI application and the GUI
      * is generating messages before the process terminated.
      */
-    while (PeekMessage(&msg, NULL, 0, 0, PM_NOREMOVE|PM_NOYIELD))
-        /* keep going */ ;
+    PeekMessage(&msg, NULL, 0, 0, PM_NOREMOVE|PM_NOYIELD);
 
     /* Above or other stuff may have set a signal flag */
     if (PL_sig_pending)

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

On 13/03/2008\, Robert May \<rob@&#8203;themayfamily\.me\.uk> wrote&#8203;:
> A problem with my changes to give win32 proper messages loops \(applied
>  as change \#33255\) has been identified on Vista\.
>
>  The patch below fixes this \(patch against blead@&#8203;33505\)\.  I haven't
>  checked it out on other win32 OS\, but is closer to what was in
>  win32\_async\_check\(\) before I got my hands on it\.  I can't claim to
>  understand the difference in behaviour on Vista to other OS\, but the
>  proposed change is more consistent with the MS documentation for
>  PeekMessage\(\)\.
>
>  I believe that with this change bug \#51674 can be closed\.

Thanks\, applied as \#33506\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
